flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,296 @@
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+
+		this.eventLoopGroup = new NioEventLoopGroup();
+		this.subscriber = pubSubSubscriberFactory.getSubscriber(eventLoopGroup, credentials);
+		this.deduplicateMessages = getRuntimeContext().getNumberOfParallelSubtasks() == 1;
+		this.isRunning = true;
+		this.pullRequest = PullRequest.newBuilder()
+										.setMaxMessages(100)
+										.setReturnImmediately(false)
+										.setSubscription(projectSubscriptionName)
+										.build();
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<String> acknowledgementIds) {
+		if (acknowledgementIds.isEmpty() || !isRunning) {
+			return;
+		}
+
+		AcknowledgeRequest acknowledgeRequest =
+			AcknowledgeRequest.newBuilder()
+								.setSubscription(projectSubscriptionName)
+								.addAllAckIds(acknowledgementIds)
+								.build();
+		subscriber.acknowledgeCallable().call(acknowledgeRequest);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		while (isRunning) {
+			messagesFuture = subscriber.pullCallable().futureCall(pullRequest);
+			try {
+				processMessage(sourceContext, messagesFuture.get().getReceivedMessagesList());
+			} catch (InterruptedException | CancellationException e) {
+				awaitSubscriberTermination();
+				return;
+			}
+		}
+
+		awaitSubscriberTermination();
+	}
+
+	void processMessage(SourceContext<OUT> sourceContext, List<ReceivedMessage> messages) throws IOException {
+		synchronized (sourceContext.getCheckpointLock()) {
+			for (ReceivedMessage message : messages) {
+				sessionIds.add(message.getAckId());
+
+				PubsubMessage pubsubMessage = message.getMessage();
+				if (deduplicateMessages && !addId(pubsubMessage.getMessageId())) {
+					// message is duplicate so just ignore it
+					return;
+				}
+
+				OUT deserializedMessage = deserializeMessage(pubsubMessage);
+				if (deserializationSchema.isEndOfStream(deserializedMessage)) {
+					stop();
+					return;
+				}
+
+				sourceContext.collect(deserializedMessage);
+			}
+
+		}
+	}
+
+	private Integer getOutstandingMessagesToAck() {
+		return this.sessionIdsPerSnapshot
+				.stream()
+				.mapToInt(tuple -> tuple.f1.size())
+				.sum() + this.sessionIds.size();
+	}
+
+	@Override
+	public void cancel() {
+		stop();
+	}
+
+	@Override
+	public void stop() {
```

Comment:
If stop will be invoked in a different thread, is it thread safe to access `messageFuture`, `subscriber` and `eventLoopGroup`? It might be better to just set isRunning to false, let the main loop pick it up then do the clean up.

Reply:
I can't find much about this in their docs. 

A few things to note:
- All of them represent background threads that are doing work
-  and all these shutdown() functions have async behavior, ie they do not block or wait for it to shutdown

This makes me think thread safety has been taken care of. Would you agree?

Reply:
Even the classes used in here are all thread safe, the thread safety is still a problem here.

For example, `messageFuture` variable is assigned different instances all the time. If `messagesFuture.cancel()` is invoked on one instance, while the main thread is calling `get()` on another instance, the main thread may block on the future for long. As a result the Source task will not exit in time.

Also, since the semantic of `stop` method is to stop gracefully. Usually we don't want the thread calling stop to execute the shutdown routine. This may complicate things. For example, what is the behavior of `subscriber.pullCallable().futureCall()` when the underlying `eventLoopGroup` has shutdown? Will it throw exception? If so, that exception is not caught in the current code and will cause the task to fail.

So a simpler and usually safer way to do the shutdown is to just set a flag and ensure the main thread knows it is being shutting down (e.g. via interrupting). The main thread will then execute the shutdown routine by itself to do the cleanup.

Reply:
Yeah this makes sense, how would I let the main thread know it should be shutting down? Specifically how can I interrupt the main thread from the `stop()` function?

Reply:
It seems the simplest way here is to call `Thread.currentThread()` in `run()` and save the thread in a variable, say `sourceThread`. When `stop()` is invoked, it will set the shutdown flag and call `sourceThread.interrupt()`.

Reply:
I've removed the StoppableFunction interface as I noticed it might get deprecated soonish in FLINK-11889 and have changed the `cancel()` to just set a boolean to false as per your suggestion.

Could you have another look?